### PR TITLE
added support for "extended local headers" in ZIP files

### PR DESCRIPTION
**Pull request by ChrisEt**
## Description

In this Pull Request I propose to add a piece of code in order to support a special format of ZIP files.

The code extends ZIP file processing in _CZipFile::UnpackFromMemory()_. Without my addition, this method just aborts with an error message if it encounters a so-called "extended local header" (also called "data descriptor record" in some sources). More details about this header can be found here: http://www.onicos.com/staff/iz/formats/zip.html. 

I modified the loop which runs through the ZIP file. If the flags indicate that an extended local header is present, I start an extra loop to find it and read its contents (this is the only way unfortunately, because the extended local header comes **after** the compressed data and their length is given nowhere). After the extended local header has been read, processing of the compressed data continues as usual.

## Motivation and Context

This method is used by several scrapers, XEM for example. It happened to me with a few TV shows that XEM failed to retrieve the show's information and instead aborted with the error message that "extended local headers are not supported".

It might be possible to reproduce the problem by using XEM to scrape information for a folder called "The Curse of Oak Island". I say "might" because this type of ZIP file doesn't occur everywhere and when this show is updated, the new ZIP file might be different.

## How Has This Been Tested?

At the time of writing this pull request, I could test my change successfully with the TV shows "The Magicians (2015)" and "The Curse of Oak Island". I compiled Kodi and created two folders with that name with an empty file in them. Using the XEM addon I could successfully scrape the information of these shows. 

Since the remote ZIP files can change anytime (and for easier debugging), I also implemented a unit test. Just run "kodi-test" and it will execute _TestZipFile.cpp_. For this test I added a sample ZIP file to Git that I fetched from thetvdb.com which contains these special headers.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

File: xbmc/filesystem/ZipFile.cpp
Comment (fritsch) [inline xbmc/filesystem/ZipFile.cpp]: Are you sure you need the == ?

Comment (ChrisEt) [reply xbmc/filesystem/ZipFile.cpp]: No, of course not :smile: I left it as it was from the original author.

File: xbmc/filesystem/ZipFile.cpp
Comment (fritsch) [inline xbmc/filesystem/ZipFile.cpp]: Shouldn 't this be "|" to compute the checksum?

Comment (ChrisEt) [reply xbmc/filesystem/ZipFile.cpp]: Yes... I know, embarrassing. Found it myself and fixed it in the second commit.
Should I have squashed my commits? (sorry, it's my first PR)

Comment (fritsch) [reply xbmc/filesystem/ZipFile.cpp]: Yes, please. Sorry - I read from top to bottom.

**Comment by fritsch** [comment]
And btw. Thanks very much for contributing to koid. Plain forgot that before my first comment. Highly appreciated.

File: xbmc/filesystem/ZipFile.cpp
Comment (ChrisEt) [reply xbmc/filesystem/ZipFile.cpp]: Will do in my next PR :thumbsup:

Be careful with the left-shifts - at least on my platform (Visual Studio) these were _signed char_. As a result, any bytes bigger than 0x80 caused problems: `(signed char)0x81 << 8 == 0xffff8100`. That's why I had add the cast to _(uint8_t)_

File: xbmc/filesystem/ZipFile.cpp
Comment (fritsch) [inline xbmc/filesystem/ZipFile.cpp]: Please use static_cast<uint8_t>(....) - this is one of our code guideline conventions

Comment (ChrisEt) [reply xbmc/filesystem/ZipFile.cpp]: Ok did it. I think my commit already got pushed into this PR, can you confirm?

**Comment by ChrisEt** [comment]
I added a new (sqashed) commit. All the review comments have been integrated, also the Jenkins tests should work now.

**Review by fritsch — APPROVED** [review]

File: xbmc/filesystem/ZipFile.cpp
Comment (Rechi) [inline xbmc/filesystem/ZipFile.cpp]: remove the white space at the end of the line

File: xbmc/filesystem/ZipFile.cpp
Comment (Rechi) [inline xbmc/filesystem/ZipFile.cpp]: > Commas have to be followed by one space.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#34-whitespace

File: xbmc/filesystem/ZipFile.cpp
Comment (Rechi) [inline xbmc/filesystem/ZipFile.cpp]: remove the white space at the end of the line

File: xbmc/filesystem/ZipFile.cpp
Comment (Rechi) [inline xbmc/filesystem/ZipFile.cpp]: one space between `;` and `//` is enough

File: xbmc/filesystem/ZipFile.cpp
Comment (Rechi) [inline xbmc/filesystem/ZipFile.cpp]: remove the white space at the end of the line

File: xbmc/filesystem/test/TestZipFile.cpp
Comment (Rechi) [inline xbmc/filesystem/test/TestZipFile.cpp]: > Included header files have to be sorted (case sensitive) alphabetically to prevent duplicates and allow better overview, with an empty line clearly separating sections.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-headers

File: xbmc/filesystem/test/TestZipFile.cpp
Comment (Rechi) [inline xbmc/filesystem/test/TestZipFile.cpp]: one space between `;` and `//` is enough

File: xbmc/filesystem/test/TestZipFile.cpp
Comment (Rechi) [inline xbmc/filesystem/test/TestZipFile.cpp]: one space between `;` and `//` is enough

File: cmake/installdata/test-reference-data.txt
Comment (Rechi) [inline cmake/installdata/test-reference-data.txt]: move this a few lines up, to keep it alphabetically sorted

**Review by Rechi — COMMENTED** [review]
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

**Comment by ChrisEt** [comment]
Integrated code style findings from @Rechi

**Comment by Sesshoumaru-sama** [comment]
Please consider for Leia if possible - error is spreading as more sources seem to switch to extended headers.

**Comment by aKzenT** [comment]
When will this be available in a release? I'm seeing lots of errors for zip files with extended local headers in my log file.

**Comment by phunkyfish** [comment]
jenkins build this please

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by phunkyfish** [comment]
Will merge today if no objections.

**Comment by phunkyfish** [comment]
@ChrisEt can you prepare the backport once merged?

**Comment by ChrisEt** [comment]
> 
> 
> @ChrisEt can you prepare the backport once merged?

Sure, I'd be happy to! For which versions will the backport be needed (sorry if that's a stupid question, this is my first pull request for xbmc)?

**Comment by phunkyfish** [comment]
No worries, I'd rather you ask the question! Just for Leis (v18). That way if/when we do a 18.7 release all users get benefit from this.

**Pull request merged**
